### PR TITLE
Basic nvrhi renderer

### PR DESCRIPTION
**Pull request by Twarug**
Allows entities to draw triangles

### CODE REVIEW

File: include/gfx/Renderer.h [hunk 1]
```diff
@@ -1,17 +1,60 @@
...
+class Mesh;
+
+class Renderer: public std::enable_shared_from_this<Renderer> {
+public:
+	static Ref<Renderer> create(RenderingAPI api);
+
+	static Ref<Renderer> current() { return s_current; }
+
+protected:
+	explicit Renderer(RenderingAPI api, bool debug = true): _api(api), _debug(debug) {}
```
Comment (Chris-plusplus) [inline include/gfx/Renderer.h:20]: Can be moved to 2nd `protected:`

File: include/gfx/Texture.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include "gfx/texture/Texture.h"
+#include "gfx/texture/TextureFilterMode.h"
+#include "gfx/texture/TextureWrapMode.h"
+#include "texture/TextureManager.h"
+
+namespace arch::gfx {
+
+using namespace arch::gfx::texture;
```
Comment (Chris-plusplus) [inline include/gfx/Texture.h:10]: Should everything from `gfx::texture` be visible from `arch`?

File: include/gfx/buffer/Buffer.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	explicit Buffer(BufferType type): _type(type) {}
+
+	virtual ~Buffer() = default;
+
+public:
+	static Ref<Buffer> create(BufferType type);
+	static Ref<Buffer> create(size_t size, BufferType type);
+	static Ref<Buffer> create(const void* data, size_t size, BufferType type);
+
+public:
```
Comment (Chris-plusplus) [inline include/gfx/buffer/Buffer.h:20]: 2 consecutive `public:`

File: include/gfx/buffer/Buffer.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#include "Mmath.h"
+#include "Ref.h"
+
+namespace arch::gfx::buffer {
+
+class Buffer {
+protected:
+	explicit Buffer(BufferType type): _type(type) {}
+
+	virtual ~Buffer() = default;
```
Comment (Chris-plusplus) [inline include/gfx/buffer/Buffer.h:13]: `Buffer` and `~Buffer` can be moved to 2nd `protected:`

File: include/gfx/buffer/BufferManager.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+#include "VertexBuffer.h"
+
+namespace arch::gfx::buffer {
+
+class BufferManager {
+public:
+	BufferManager() = default;
+	virtual ~BufferManager() = default;
+
+public:
```
Comment (Chris-plusplus) [inline include/gfx/buffer/BufferManager.h:16]: 2 consecutive `public:`

File: include/gfx/Mesh.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+public:
+	Mesh(const Ref<buffer::VertexBuffer>& vertices, const Ref<buffer::IndexBuffer>& indices);
+	~Mesh();
+
+public:
+	template<typename VertexT>
+	static Ref<Mesh> create(std::span<VertexT> vertices, std::span<u32> indices);
+	static Ref<Mesh> create(Ref<buffer::VertexBuffer> vertices, Ref<buffer::IndexBuffer> indices);
+
+public:
```
Comment (Chris-plusplus) [inline include/gfx/Mesh.h:23]: 3 consecutive `public:`

File: include/gfx/texture/Texture.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	GraphicsFormat getFormat() const;
+
+	bool isReadable() const;
+
+	u32 getWidth() const;
+	u32 getHeight() const;
+	u32 getDepth() const;
+	u32 getSize(u32 axis) const;
+
+public:
```
Comment (Chris-plusplus) [inline include/gfx/texture/Texture.h:33]: 2 consecutive `public:`

File: include/gfx/texture/TextureManager.h [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+#include <filesystem>
+
+#include "Ref.h"
+#include "Texture.h"
+
+namespace arch::gfx::texture {
+
+class TextureManager {
+protected:
+	TextureManager() = default;
```
Comment (Chris-plusplus) [inline include/gfx/texture/TextureManager.h:12]: can be moved to 2nd `protected:`

File: include/platform/nvrhi/NvrhiContext.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+	virtual void onResize(u32 width, u32 height) = 0;
+
+	virtual void beginFrame() = 0;
+	virtual void present() = 0;
+
+public:
+	::nvrhi::FramebufferHandle getFramebuffer(int index);
+
+public:
```
Comment (Chris-plusplus) [inline include/platform/nvrhi/NvrhiContext.h:29]: 3 consecutive `public:`

File: include/platform/nvrhi/NvrhiContext.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+protected:
+	void _preResizeFramebuffers();
+	void _postResizeFramebuffers();
+
+protected:
+	virtual u32 _getBackBufferCount() const = 0;
+	virtual ::nvrhi::TextureHandle _getBackBuffer(uint32_t index) const = 0;
+
+protected:
```
Comment (Chris-plusplus) [inline include/platform/nvrhi/NvrhiContext.h:43]: 3 consecutive `protected:`

File: include/platform/nvrhi/NvrhiRenderer.h [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	void beginFrame() override;
+	void present() override;
+
+	void render(const Ref<Mesh>& mesh, const Mat4x4& transform) override;
+
+public:
+	Ref<gfx::buffer::BufferManager> getBufferManager() override;
+	Ref<gfx::texture::TextureManager> getTextureManager() override;
+
+public:
```
Comment (Chris-plusplus) [inline include/platform/nvrhi/NvrhiRenderer.h:36]: 4 consecutive `public:`

File: include/platform/nvrhi/NvrhiUtils.h [hunk 12]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+#include "gfx/GraphicsFormat.h"
+#include "nvrhi/nvrhi.h"
+
+namespace arch::gfx::nvrhi {
+
+class NvrhiUtils {
+public:
+	static ::nvrhi::Format getFormat(GraphicsFormat format);
```
Comment (Chris-plusplus) [inline include/platform/nvrhi/NvrhiUtils.h:9]: `NvrhiUtils` could be made into namespace

File: include/platform/nvrhi/buffer/NvrhiBuffer.h [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+#include "nvrhi/nvrhi.h"
+
+namespace arch::gfx::nvrhi::buffer {
+
+class NvrhiBuffer: public virtual gfx::buffer::Buffer {
+public:
+	NvrhiBuffer(gfx::buffer::BufferType type);
+	~NvrhiBuffer() override;
+
+public:
```
Comment (Chris-plusplus) [inline include/platform/nvrhi/buffer/NvrhiBuffer.h:12]: 2 consecutive `public:`

File: include/platform/nvrhi/context/NvrhiVulkanContext.h [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	void beginFrame() override;
+	void present() override;
+
+public:
+	int getCurrentFrameIndex() override;
+	uint2 getFramebufferSize() const override;
+
+	::nvrhi::DeviceHandle getDevice() override;
+
+public:
```
Comment (Chris-plusplus) [inline include/platform/nvrhi/context/NvrhiVulkanContext.h:30]: 3 consecutive `public:`

File: include/platform/nvrhi/context/NvrhiVulkanContext.h [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+private:
+	void _createSurface(const Ref<Window>& window);
+	void _createSwapchain(i32 width, i32 height);
+	void _destroySwapchain();
+	void _createFrameSemaphores();
+	void _destroyFrameSemaphores();
+
+public:
+
+private:
```
Comment (Chris-plusplus) [inline include/platform/nvrhi/context/NvrhiVulkanContext.h:49]: emtpy `public:`

Comment (Chris-plusplus) [reply include/platform/nvrhi/context/NvrhiVulkanContext.h:49]: both `private:` could be merged

File: include/platform/vulkan/VulkanCommandBuffer.h
Comment (Chris-plusplus) [inline include/platform/vulkan/VulkanCommandBuffer.h:1]: empty file

File: include/platform/vulkan/VulkanContext.h [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+	VulkanContext(const VulkanContext&) = delete;
+	VulkanContext& operator=(const VulkanContext&) = delete;
+
+	VulkanContext(VulkanContext&&) = default;
+	VulkanContext& operator=(VulkanContext&&) = default;
+
+public:
+	void initDevice(VkSurfaceKHR surface);
+
+public:
```
Comment (Chris-plusplus) [inline include/platform/vulkan/VulkanContext.h:57]: 4 consecutive `public:`

File: include/platform/vulkan/VulkanContext.h [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+	const Queue& getQueue(QueueType type) const;
+
+	i32 findMemoryType(u32 typeFilter, VkMemoryPropertyFlags properties) const;
+	VkFormat findSupportedFormat(
+		const std::vector<VkFormat>& candidates,
+		VkImageTiling tiling,
+		VkFormatFeatureFlags features
+	) const;
+
+private:
```
Comment (Chris-plusplus) [inline include/platform/vulkan/VulkanContext.h:77]: can be merged with 2nd `private:`

File: include/platform/vulkan/VulkanRenderer.h [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+	void _createDepthTexture();
+	void _createFrames();
+
+public:
+	void onResize(u32 width, u32 height) override;
+
+	void beginFrame() override;
+	void present() override;
+
+public:
```
Comment (Chris-plusplus) [inline include/platform/vulkan/VulkanRenderer.h:44]: 2 consecutive `public:`

File: include/platform/vulkan/VulkanRenderer.h [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+	Ref<buffer::VulkanBufferManager> _bufferManager;
+	Ref<texture::VulkanTextureManager> _textureManager;
```
Comment (Chris-plusplus) [inline include/platform/vulkan/VulkanRenderer.h:46]: public members with _ at the beginning

File: include/platform/vulkan/VulkanRenderer.h [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+		VkFramebuffer framebuffer;
+
+		VkCommandBuffer commandBuffer;
+		VkSemaphore imageAvailableSemaphore;
+		VkSemaphore renderFinishedSemaphore;
+		VkFence inFlightFence;
+	};
+
+	std::vector<FrameData> frames;
+	std::uint32_t _frameIndex = 0;
```
Comment (Chris-plusplus) [inline include/platform/vulkan/VulkanRenderer.h:61]: public member with _ at the beginning

File: include/platform/vulkan/VulkanSwapchain.h [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+	VulkanSwapchain(const VulkanSwapchain&) = delete;
+	VulkanSwapchain& operator=(const VulkanSwapchain&) = delete;
+
+	VulkanSwapchain(VulkanSwapchain&&) = default;
+	VulkanSwapchain& operator=(VulkanSwapchain&&) = default;
+
+public:
+	void updateSwapchain();
+
+public:
```
Comment (Chris-plusplus) [inline include/platform/vulkan/VulkanSwapchain.h:44]: 4 consecutive `public:`

File: include/platform/vulkan/VulkanSwapchain.h [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+	const Frame& getFrame(int index) const { return _frames[index]; }
+
+private:
+	void _createSurface();
+
+	void _createSwapchain();
+	void _cleanupSwapchain();
+	void _recreateSwapchain();
+
+private:
```
Comment (Chris-plusplus) [inline include/platform/vulkan/VulkanSwapchain.h:64]: 2 consecutive `private:`

File: include/platform/vulkan/buffer/VulkanBufferManager.h [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+public:
+	explicit VulkanBufferManager(const Ref<VulkanContext>& context);
+	~VulkanBufferManager() noexcept override = default;
+
+private:
+	Ref<VertexBuffer> _createVertexBufferImpl(void* data, u32 size) override;
+	Ref<IndexBuffer> _createIndexBufferImpl(std::span<u32> indices) override;
+	Ref<Buffer> _createBufferImpl(void* data, u32 size) override;
+
+private:
```
Comment (Chris-plusplus) [inline include/platform/vulkan/buffer/VulkanBufferManager.h:18]: 2 consecutive `private:`

File: include/platform/vulkan/texture/VulkanImage.h [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+		VkMemoryPropertyFlags memoryFlags
+	);
+	~VulkanImage();
+
+	void transitionImageLayout(VkImageLayout oldLayout, VkImageLayout newLayout);
+	void createImageView(const Ref<VulkanContext>& context, u16 layerCount, i32 layerIndex, VkFormat format);
+
+	void cleanup(const Ref<VulkanContext>& context);
+
+public:
```
Comment (Chris-plusplus) [inline include/platform/vulkan/texture/VulkanImage.h:34]: 2 consecutive `public:`

File: include/scene/SceneManager.h [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+	void update();
+	void renderScene(const Ref<gfx::Renderer>& renderer);
+
+	const Ref<Scene>& currentScene() const noexcept;
+	void changeScene(const Ref<Scene>& scene);
+
+private:
+	Ref<Scene> _currentScene;
+
+public:
```
Comment (Chris-plusplus) [inline include/scene/SceneManager.h:25]: can be merged with 1st `public:`

File: include/scene/components/TransformComponent.h
Comment (Chris-plusplus) [inline include/scene/components/TransformComponent.h:1]: why make setters & getters? fields are public

File: src/platform/nvrhi/context/NvrhiVulkanContext.cpp [hunk 26]
```diff
@@ -0,0 +1,371 @@
...
+#include "nvrhi/validation.h"
+#include "nvrhi/vulkan.h"
+#include "platform/nvrhi/NvrhiMessageCallback.h"
+#include "platform/nvrhi/NvrhiUtils.h"
+#include "platform/vulkan/VulkanUtils.h"
+
+VULKAN_HPP_DEFAULT_DISPATCH_LOADER_DYNAMIC_STORAGE
+
+#undef min
+#undef max
```
Comment (Chris-plusplus) [inline src/platform/nvrhi/context/NvrhiVulkanContext.cpp:15]: (optional) can be replaced with `#define NOMINMAX` at the beginning of file

File: archimedes_bin/main.cpp
Comment (Chris-plusplus) [inline archimedes_bin/main.cpp:1]: (optional) could be moved into dedicated trianglesExample or something

**Review by Chris-plusplus — CHANGES_REQUESTED** [review]

File: include/gfx/Buffer.h [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+
+#include "buffer/Buffer.h"
+#include "buffer/BufferManager.h"
+#include "buffer/BufferType.h"
+#include "buffer/IndexBuffer.h"
+#include "buffer/VertexBuffer.h"
+
+namespace arch::gfx {
+
+using namespace arch::gfx::buffer;
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/gfx/Buffer.h:11]: Gfx.h has:
```cpp
namespace arch {
    using namespace arch::gfx;
}
```
so everything from `gfx` (`buffer` included) is in `arch::`
image

File: include/gfx/Renderer.h (hunk 1 above)
Comment (Twarug) [reply include/gfx/Renderer.h:20]: These help divide API to logical parts.

File: include/gfx/Renderer.h [hunk 28]
```diff
@@ -1,17 +1,60 @@
...
+	virtual void onResize(u32 width, u32 height) = 0;
+
+	virtual void beginFrame() = 0;
+	virtual void present() = 0;
+
+public:
+
+	virtual void render(const Ref<Mesh>& mesh, const Mat4x4& transform) = 0;
+
+public:
```
[the remark this answers is not in the document]
Comment (Twarug) [reply include/gfx/Renderer.h:45]: These help divide API to logical parts.

File: include/scene/SceneManager.h (hunk 25 above)
Comment (Twarug) [reply include/scene/SceneManager.h:25]: These help divide API to logical parts.

File: include/gfx/Buffer.h (hunk 27 above)
[the remark this answers is not in the document]
Comment (Twarug) [reply include/gfx/Buffer.h:11]: yup, but `arch::gfx` no longer will be pulled to `arch` namespace

File: include/gfx/Texture.h (hunk 2 above)
Comment (Twarug) [reply include/gfx/Texture.h:10]: no, will be visible in `arch::gfx` tho

**Review by Chris-plusplus — APPROVED** [review]

**Pull request merged**